File: 0155-min-stack/0155-min-stack.c



typedef struct mstack {
    int val;
    int min;
    struct mstack* next;
} MinStack;


MinStack* minStackCreate() {
    MinStack* obj = malloc(sizeof(MinStack));
    obj->next = obj;
    return obj;
}

void minStackPush(MinStack* obj, int val) {
    if (obj->next == obj){
        obj->val = val;
        obj->min = val;
        obj->next = NULL;
    } else {
        MinStack* newScnd = malloc(sizeof(MinStack));
        newScnd->val = obj->val;
        newScnd->min = obj->min;
        newScnd->next = obj->next;
        obj->next = newScnd;
        obj->val = val;
        obj->min = (val < newScnd->min) ? val : newScnd->min;
        printf("%d ", val);
    }
  
}

void minStackPop(MinStack* obj) {
    if (obj->next != obj && obj->next != NULL) {
        MinStack *killme = obj->next;
        obj->val = obj->next->val;
        obj->min = obj->next->min;
        obj->next = obj->next->next;
        free(killme);
    } else {
        obj->next = obj;
    }
}

int minStackTop(MinStack* obj) {
    return (obj == NULL) ? 0 : obj->val;
}

int minStackGetMin(MinStack* obj) {
    return (obj == NULL) ? 0 : obj->min;
}

void minStackFree(MinStack* obj) {
    while (obj != obj->next){
        minStackPop(obj);
    }
}

/**
 * Your MinStack struct will be instantiated and called as such:
 * MinStack* obj = minStackCreate();
 * minStackPush(obj, val);
 
 * minStackPop(obj);
 
 * int param_3 = minStackTop(obj);
 
 * int param_4 = minStackGetMin(obj);
 
 * minStackFree(obj);
*/

File: 0155-min-stack/0155-min-stack.py
class MinStack(object):

    def __init__(self):
        # values are pairs (value, min) where min is the minimum
        # at the moment value was added
        self.values = []
        

    def push(self, val):
        """
        :type val: int
        :rtype: None
        """
        newMin = min(val, self.values[-1][1]) if self.values else val
        self.values.append((val, newMin))

    def pop(self):
        """
        :rtype: None
        """
        if self.values:
            self.values.pop(-1)

        if self.values:
            self.min = self.values[-1][1]

    def top(self):
        """
        :rtype: int
        """
        return self.values[-1][0] if self.values else None

    def getMin(self):
        """
        :rtype: int
        """
        return self.values[-1][1] if self.values else None
        


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()